File: src/main/java/at/searles/buf/ReaderCharStream.java
package at.searles.buf;

import java.io.IOException;
import java.io.Reader;

/**
 * This class creates a CharStream out of a Reader.
 */
public class ReaderCharStream implements CharStream {
    private final Reader r;
    private long offset;

    public ReaderCharStream(Reader r) {
        this.r = r;
        this.offset = 0;
    }

    @Override
    public int next() {
        try {
            int ch = r.read();

            if(ch == -1) {
                return -1;
            }

            offset ++;

            if(Character.isHighSurrogate((char) ch)) {
                int low = r.read();

                if(low == -1) {
                    throw new IllegalArgumentException("no lo surrogate from this reader");
                }

                offset ++;

                return Character.toCodePoint((char) ch, (char) low);
            }

            return ch; // char = codepoint.
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }

    @Override
    public String toString() {
        return r.toString() + "@" + offset;
    }
}

File: src/main/java/at/searles/buf/ReaderCharStream.kt
package at.searles.buf

import java.io.IOException
import java.io.Reader

/**
 * This class creates a CharStream out of a Reader.
 */
class ReaderCharStream(private val r: Reader) : CharStream {

    private var offset: Long = 0

    override fun next(): Int {
        return try {
            val ch = r.read()
            if (ch == -1) {
                return -1
            }
            offset++
            if (Character.isHighSurrogate(ch.toChar())) {
                val low = r.read()
                require(low != -1) { "no lo surrogate from this reader" }
                offset++
                return Character.toCodePoint(ch.toChar(), low.toChar())
            }
            ch // char = codepoint.
        } catch (e: IOException) {
            throw IllegalStateException(e)
        }
    }

    override fun toString(): String {
        return "$r@$offset"
    }

}